sdk/python/kfp/compiler/compiler_utils.py

```diff
@@ -427,14 +428,42 @@ def get_dependencies(
                 task2=task,
             )
 
-            # a task cannot depend on a task created in a for loop group since individual PipelineTask variables are reassigned after each loop iteration
-            dependent_group = group_name_to_group.get(upstream_groups[0], None)
-            if isinstance(dependent_group,
-                          (tasks_group.ParallelFor, tasks_group.Condition,
-                           tasks_group.ExitHandler)):
+            # uncommon upstream ancestor check
+            uncommon_upstream_groups = deepcopy(upstream_groups)
+            uncommon_upstream_groups.remove(
+                upstream_task.name
+            )  # because a task's `upstream_groups` contains the task's name
+            if uncommon_upstream_groups:
+                dependent_group = group_name_to_group.get(
+                    uncommon_upstream_groups[0], None)
+                if isinstance(dependent_group, tasks_group.ExitHandler):
+                    task_group_type = 'an ' + tasks_group.ExitHandler.__name__
+
+                elif isinstance(dependent_group, tasks_group.Condition):
+                    task_group_type = 'a ' + tasks_group.Condition.__name__
+
+                else:
+                    task_group_type = 'a ' + tasks_group.ParallelFor.__name__
+
                 raise RuntimeError(
```

Comment:
No action required / not your code: I want to note here that I'm not sure if this should use a `RuntimeError`.  This _is_ a runtime error in the sense that it results an ambiguous runtime topology, but it's not a true "runtime" error at the time it's raised.

This file used `RuntimeError` before this PR and, since this PR actually reduces the set of topologies for which this error would be raised, we don't necessarily need to reconsider this in this PR.

Furthermore, an `Exception` is usually used when the error is attributed to user code, whereas an `Error` is usually used when the error is attributed to something else, such as an environment. In this case, this is user code.

For this reason, I think it would make sense for this to be a custom `InvalidTopologyException` or something similar.

Relatedly, some of the `ValueErrors` from `pipeline_task.py` now become `RuntimeErrors` in this PR, so perhaps that is a reason to consider this in the short term.